### PR TITLE
Add Docker (Compose) support

### PR DESCRIPTION
**Pull request by easybe**
A first stab at #31 and #54.

To do:
- [x] Look into CI integration
- [x] Configure linter for YAML (the whole world uses 2 spaces for Docker Compose files)
- [x] Add service user to the image, it is bad practice to run the application as root
- [x] Use `gunicorn`
- [x] Make sure we can run behind a reverse proxy (NGINX)

---

This can be tested with:
```
docker compose up --build
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+FROM python:3.12.4-alpine
+
+COPY . /tmp/src
+RUN pip install --no-cache-dir /tmp/src && \
```
Comment (easybe) [inline Dockerfile:4]: For the image that we upload to the registry I would rather install the package from PyPI here.

File: pyproject.toml [hunk 2]
```diff
@@ -24,6 +24,12 @@ opentelemetry-instrumentation-fastapi = "^0.47b0"
 opentelemetry-exporter-prometheus = "^0.47b0"
 aiocache = "^0.12.2"
 httpx = "^0.27.0"
+dynaconf = "^3.2.6"
+
+asyncpg = { version = "^0.29.0", optional = true }
+
+[tool.poetry.extras]
+postgresql = ["asyncpg"]
```
Comment (easybe) [inline pyproject.toml:32]: I could not figure out how to reference the `tortoise-orm[asyncpg]` extra here.

Comment (b-rowan) [reply pyproject.toml:32]: I think this is fine for now.

File: goosebit/settings.py [hunk 3]
```diff
@@ -2,23 +2,24 @@
...
 PWD_CXT = PasswordHasher()
 
-with open(BASE_DIR.joinpath("settings.yaml"), "r") as f:
-    config = yaml.safe_load(f.read())
+config = Dynaconf(
+    envvar_prefix="GOOSEBIT",
+    settings_files=[BASE_DIR.joinpath("settings.yaml"), BASE_DIR.joinpath(".secrets.yaml")],
+)
+
+UPDATES_DIR = Path(config.get("artifacts_dir", BASE_DIR.joinpath("updates")))
```
Comment (b-rowan) [inline goosebit/settings.py:22]: Might as well rename this from `UPDATES_DIR` to `ARTIFACTS_DIR`.

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    image: goosebit
+    build:
+      context: .
+      args:
+        GOOSEBIT_VERSION: 0.1.2
+    environment:
+      GOOSEBIT_DB_URI: postgres://postgres:admin@postgres:5432/goosebit
+      GOOSEBIT_USERS: >-
+        [{email="admin@goosebit.local", password="admin", permissions="*"}]
+      GOOSEBIT_ARTIFACTS_DIR: /artifacts
```
Comment (b-rowan) [inline docker-compose.yml:27]: Can you confirm that these settings get imported properly?  I won't be able to test this for a bit yet, so I just want to confirm these actually get loaded properly.

**Review by b-rowan — APPROVED** [review]
Looks good overall, just a couple housekeeping/clarification things.

**Comment by b-rowan** [comment]
Regarding CI, I already have an action set up for version releasing, I think we could add a build and push to docker there.

The only concerns I have is we have to figure out how to update the version in the docker compose.

**Comment by b-rowan** [comment]
> Regarding CI, I already have an action set up for version releasing, I think we could add a build and push to docker there.
> 
> The only concerns I have is we have to figure out how to update the version in the docker compose.

Here is a GitHub guide on doing it using actions.  I have added a `DOCKER_USERNAME` and `DOCKER_PASSWORD` value to the secrets for the repo, so those should be usable.  This will push to the `upstreamdata` docker page on docker hub.

File: docker-compose.yml (hunk 4 above)
Comment (easybe) [reply docker-compose.yml:27]: Yes, everything works.

**Comment by easybe** [comment]
> Regarding CI, I already have an action set up for version releasing, I think we could add a build and push to docker there.
> 
> The only concerns I have is we have to figure out how to update the version in the docker compose.

Yes, that is a tricky one. It will be hard to avoid having the version in 2 places. However, the Compose file shall only be used as an example or for development. Build and push I plan to do with GitHub Actions directly.

File: goosebit/settings.py (hunk 3 above)
Comment (easybe) [reply goosebit/settings.py:22]: I would then also rename the `updates` directory to `artifacts`. I like the name better, as the files don't necessarily have to be updates. Is that an option for you?

Comment (b-rowan) [reply goosebit/settings.py:22]: > I would then also rename the `updates` directory to `artifacts`. I like the name better, as the files don't necessarily have to be updates. Is that an option for you?

:100: this is much better.

**Comment by b-rowan** [comment]
>  Configure linter for YAML (the whole world uses 2 spaces for Docker Compose files)

I generally like to keep things consistent across the repo, meaning 4 spaces indentation for all files.  I'm not strictly opposed to this for yaml files, but something we should do some deliberation about.

For reference, this is handled by `prettier`, which does both HTML and YAML files.  I want to keep HTML at 4 spaces, but if we wanted to handle YAML differently we would have to figure out how to in `.prettierrc.yaml`.

**Comment by b-rowan** [comment]
> > Configure linter for YAML (the whole world uses 2 spaces for Docker Compose files)
> 
> I generally like to keep things consistent across the repo, meaning 4 spaces indentation for all files. I'm not strictly opposed to this for yaml files, but something we should do some deliberation about.
> 
> For reference, this is handled by `prettier`, which does both HTML and YAML files. I want to keep HTML at 4 spaces, but if we wanted to handle YAML differently we would have to figure out how to in `.prettierrc.yaml`.

https://stackoverflow.com/questions/56464727/different-prettier-rules-for-different-file-types

**Comment by easybe** [comment]
> > Configure linter for YAML (the whole world uses 2 spaces for Docker Compose files)
> 
> I generally like to keep things consistent across the repo, meaning 4 spaces indentation for all files. I'm not strictly opposed to this for yaml files, but something we should do some deliberation about.
> 
> For reference, this is handled by `prettier`, which does both HTML and YAML files. I want to keep HTML at 4 spaces, but if we wanted to handle YAML differently we would have to figure out how to in `.prettierrc.yaml`.

I get your point. I generally also use 4 spaces. However, as pretty much all YAML out there uses 2 spaces, my eyes got used to it and it makes copying & pasting from the internet easier.

**Comment by b-rowan** [comment]
> I get your point. I generally also use 4 spaces. However, as pretty much all YAML out there is indented with 2 spaces, my eyes got used to it, but more importantly, it makes copying & pasting from the internet easier.

Ok, let's move forward with 2 space yaml then.  I think that link I provided should be enough info to configure prettier, though it will need to be converted into yaml.

**Comment by easybe** [comment]
> > I get your point. I generally also use 4 spaces. However, as pretty much all YAML out there is indented with 2 spaces, my eyes got used to it, but more importantly, it makes copying & pasting from the internet easier.
> 
> Ok, let's move forward with 2 space yaml then. I think that link I provided should be enough info to configure prettier, though it will need to be converted into yaml.

#59

**Comment by easybe** [comment]
Concerning CI, what I got so far seems to work: https://github.com/UpstreamDataInc/goosebit/actions/runs/10322906355

To test the actual push to the registry is a little bit of a challenge though, as the workflow is only meant to be run on release.

I will come up with a strategy and start getting things ready for integration.

**Comment by b-rowan** [comment]
If you only want to run on release, I use this file for release to push to PyPi- https://github.com/UpstreamDataInc/goosebit/blob/master/.github/workflows/python-publish.yml

**Comment by easybe** [comment]
@b-rowan I would assume that the [upstreamdata/goosebit](https://hub.docker.com/r/upstreamdata/goosebit) repository needs to be manually created on Docker Hub.

**Comment by b-rowan** [comment]
Ok, let me try doing that, one moment.

File: .github/workflows/docker.yml [hunk 5]
```diff
@@ -0,0 +1,63 @@
+name: Docker
+
+# on:
+#   workflow_run:
+#     workflows: ["PyPI"]
+#     types:
+#       - completed
```
Comment (easybe) [inline .github/workflows/docker.yml:7]: > If you only want to run on release, I use this file for release to push to PyPi- https://github.com/UpstreamDataInc/goosebit/blob/master/.github/workflows/python-publish.yml

Yes, I know. For that, I prepared this.

**Comment by b-rowan** [comment]
Ok, repo is up on docker hub :+1:

**Comment by easybe** [comment]
I will split this into multiple PRs. First one: 60

**Comment by b-rowan** [comment]
> Use gunicorn

Here's a reference from the FastAPI docs on this, basically just run with uvicorn workers.

https://fastapi.tiangolo.com/deployment/server-workers/#gunicorn-with-uvicorn-workers

**Comment by easybe** [comment]
> > Use gunicorn
> 
> Here's a reference from the FastAPI docs on this, basically just run with uvicorn workers.
> 
> https://fastapi.tiangolo.com/deployment/server-workers/#gunicorn-with-uvicorn-workers

Yes, that is what am working. Unfortunately, I am still struggling with getting the logger to print the format I want (and lack of time as I am also working on our internal deployment). But, I am on it. My WIP branch can be found here: https://github.com/husqvarnagroup/goosebit/tree/gardena/eb/docker_compose

**Comment by b-rowan** [comment]
> > > Use gunicorn
> 
> > 
> 
> > Here's a reference from the FastAPI docs on this, basically just run with uvicorn workers.
> 
> > 
> 
> > https://fastapi.tiangolo.com/deployment/server-workers/#gunicorn-with-uvicorn-workers
> 
> 
> 
> Yes, that is what am working on. Unfortunately, I am still struggling with getting the logger to print the format I want (and lack of time as I am also working on our internal deployment). But, I am on it. My WIP branch can be found here: https://github.com/husqvarnagroup/goosebit/tree/gardena/eb/docker_compose

All good, just wanted to help out, as I figured y'all have been busy.

**Comment by easybe** [comment]
Closing in favor of #88 and #109

**Pull request closed without merging**